### PR TITLE
Added missing support for tracers in the parser

### PR DESCRIPTION
**Pull request by trinemykk**

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by andlaus** [comment]
since I'm partially to blame for this, a note: the TBLK* keyword family should actually be grid properties. this is currently not possible, so the simulator needs to access this information via the "bare metal" deck. In particular this means that modifying these keywords via ADD, MULTIPLY, etc. is currently unsupported. the same applies to the TVDP* keyword family, but with tables instead of grid properties.

**Comment by joakim-hove** [comment]
>the TBLK* keyword family should actually be grid properties. this is currently not possible,

Why is that - is it because of the dynamic naming scheme?

**Comment by andlaus** [comment]
yes. same for TVDP*.

File: opm/parser/eclipse/EclipseState/Schedule/WellTracerProperties.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#ifndef WELLTRACERPROPERTIES_HPP_HEADER_INCLUDED
+#define WELLTRACERPROPERTIES_HPP_HEADER_INCLUDED
+#include <vector>
+#include <string>
+#include <map>
+
+namespace Opm {
+
+    struct WellTracerProperties {
+        std::map< std::string, double > m_tracerProperties;
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/WellTracerProperties.hpp]: As long as the data container is accessed through `get` and `set` methods I would suggest `WellTracerProperties` was made into a class, and that the `m_tracerProperties` member was made private.

You have called the member `tracerProperties` but it seems it only holds concentrations? I.e. maybe:
```
std::map<std::string,double> tracer_concentrations;
```
would be a better choice?

File: src/opm/parser/eclipse/EclipseState/Schedule/Well.cpp [hunk 2]
```diff
@@ -208,6 +209,11 @@ namespace Opm {
         return m_solventFraction.update(timeStep, fraction);
     }
 
+    bool Well::setTracerProperties(size_t timeStep , const WellTracerProperties& newProperties) {
+        m_isProducer.update(timeStep , false);
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Well.cpp]: This is slightly schizofrenic - at an outer scope you throw an exception if the well in a question is a producer, whereas here you just update the status become an injector and go on? I would suggest only having the test in the `setTracerProperties()` - and let that throw?

File: src/opm/parser/eclipse/EclipseState/Schedule/WellTracerProperties.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            return false;
+
+    }
+
+    void WellTracerProperties::setConcentration(const std::string& name, const double& concentration) {
+        m_tracerProperties[name] = concentration;
+    }
+
+    double WellTracerProperties::getConcentration(const std::string& name) const {
+
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/WellTracerProperties.cpp]: Extra space?

File: tests/parser/WellTracerTests.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+
+BOOST_AUTO_TEST_CASE(TestDynamicWTRACER) {
+    auto deck = createDeckWithDynamicWTRACER();
+    EclipseGrid grid(10,10,10);
+    TableManager table ( deck );
+    Eclipse3DProperties eclipseProperties ( deck , table, grid);
+    Schedule schedule(deck, grid , eclipseProperties, Phases(true, true, true) , ParseContext());
+    BOOST_CHECK(deck.hasKeyword("WTRACER"));
```
Comment (joakim-hove) [inline tests/parser/WellTracerTests.cpp]: Very good that you have written tests - but I miss testing of the sorry-path; as I see it that is actually the most important testing. You have introduced a new `throw ...` statement for the situation where you try to update the tracer properties of a producer, that should be tested with something like:
```
auto deck = createDeckWithTracerInProducer();
...
BOOST_REQUIRE_THROW( Schedule(deck, grid, ecl, ..), std::invalid_argument);
```

**Comment by joakim-hove** [comment]
OK - sorry this took so long to review. When you have addressed the comments - or argued why I'm out on a hike - this is ready to be merged.

File: tests/parser/WellTracerTests.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+
+
+BOOST_AUTO_TEST_CASE(TestNoTracer) {
+    auto deck = createDeckWithOutTracer();
+    EclipseGrid grid(10,10,10);
+    TableManager table ( deck );
+    Eclipse3DProperties eclipseProperties ( deck , table, grid);
+    Schedule schedule(deck, grid , eclipseProperties, Phases(true, true, true) , ParseContext());
```
[the remark this answers is not in the document]
Comment (trinemykk) [reply tests/parser/WellTracerTests.cpp]: Thank you. This is fixed.

File: opm/parser/eclipse/EclipseState/Schedule/WellTracerProperties.hpp (hunk 1 above)
Comment (trinemykk) [reply opm/parser/eclipse/EclipseState/Schedule/WellTracerProperties.hpp]: Good suggestions, they are implemented.

File: src/opm/parser/eclipse/EclipseState/Schedule/Well.cpp (hunk 2 above)
Comment (trinemykk) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well.cpp]: Hopefully less schizofrenic after moving the throw.

File: tests/parser/WellTracerTests.cpp (hunk 4 above)
Comment (trinemykk) [reply tests/parser/WellTracerTests.cpp]: Thank you, this is now done.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Thank you for the fixes - I am ready to merge this when Jenkins is happy.

**Comment by joakim-hove** [comment]
jenkins build this please

**Pull request merged**